### DISCUSSION TITLE
Is there a way to load multiple wasm instances on the same block of memory?

### DISCUSSION THREAD

**Discussion by frank-pian** [emscripten-core/emscripten#19208] [Q&A]
I have multiple wasm instances working on multiple web works. I want them to share the same Shared Array Buffer memory.
Using the WAT build code someone has implemented similar functionality, like this:
https://developer.mozilla.org/en-US/docs/WebAssembly/Understanding_the_text_format#mutating_tables_and_dynamic_linking
https://blog.scottlogic.com/2019/07/15/multithreaded-webassembly.html

Loading multiple wasm instances in the same linear memory requires rational stack space planning. So when I was looking into the wasm work api, I found some ways to customize the location and size of the stack.
https://github.com/emscripten-core/emscripten/blob/main/system/lib/compiler-rt/stack_limits.S

So is there a way for emscripten to do this?

**Comment by frank-pian** [comment]
I've seen some discussions with similar thoughts to mine:
https://github.com/emscripten-core/emscripten/discussions/15873
https://github.com/emscripten-core/emscripten/discussions/18920

pthread and wasm_works, neither of which can satisfy our multi-threaded scenario.
In fact my goal is to be able to transfer large amounts of data quickly on different wasm modules without a lot of code changes

**Reply by sbc100** [reply]
Emscripten doesn't directly support this (running multiple separate programs in the same address space), but it does have some features that might help.    

The first thing to try would be to simply use dynamic linking.   i.e. to build your separate programs as side modules.  See https://emscripten.org/docs/compiling/Dynamic-Linking.html.  In this mode emscripten takes case of giving each module its own static data region.   This is well tested and pretty easy to use.

If that doesn't work for you, you would try building something your self by building with `-sRELOCATABLE` which produces a program which doesn't hardcode any of its locations in memory.   This would be a lot more work for you since it would be up to you to provide things like `__memory_base` and `__stack_pointer` for each of the modules yourself.   This mode is really an internal detail of the dynamic linking support but you could try to use it directly if you like.

**Reply by sbc100** [reply]
Can you say more about why neither pthreads nor wasm workers work for you?  What features are you looking for that they lack?

**Reply by frank-pian** [reply]
> Can you say more about why neither pthreads nor wasm workers work for you? What features are you looking for that they lack?

Thank you very much for your answer. Let me try to describe my needs from the beginning.
We have a large web-based application. There are wasm modules with different functions running on different web workers. Most of their work is unrelated, but some data is exchanged, so we built a bus with postmessage to transfer data to each other. As our products become more and more loaded and the amount of data gets larger, postmessage has created a bottleneck and we want to implement data transfer directly through Shared Array Buffer. 
Switching to Pthread and wasm_worker to control threads would both require significant code changes, and we don't want multiple separate wasm modules merged into one.

**Reply by frank-pian** [reply]
I see that the mutiple memories proposal has come up. 
In the future, is it possible to implement a wasm module to allocate two linear memories. One for stack allocation and the other for data interaction between modules. This way no conflicts will arise.

**Reply by sbc100** [reply]
Sadly that is not possible with a language like C/C++ where there is s single unified address space for all memory: i.e. pointer to the heap are indistinguishable from pointers to the stack, or pointers to static data, and loads from all them go through the same, single memory.
